vllm/entrypoints/openai/api_server.py

```diff
@@ -89,6 +91,34 @@ async def health() -> Response:
     await openai_serving_chat.engine.check_health()
     return Response(status_code=200)
 
+@router.get(
+    "/liveness",
+    response_model=LivenessResponse,
+    name="liveness",
+    tags=["technical"],
+)
+async def get_liveness() -> LivenessResponse:
+    """Liveness probe for k8s"""
+    liveness_msg = LivenessResponse(alive="ok")
+    return liveness_msg
+
+
+@router.get(
+    "/readiness",
+    response_model=ReadinessResponse,
+    name="readiness",
+    tags=["technical"],
+)
+async def get_readiness() -> ReadinessResponse:
+    """Readiness probe for k8s"""
+    d_worker = openai_serving_chat.engine.engine.model_executor.driver_worker
+    model_weights = d_worker.model_runner.model_memory_usage
+
+    if model_weights > 0:
+        return ReadinessResponse(ready="ok")
+    else:
+        return ReadinessResponse(ready="ko")
```

Comment:
> > @mfournioux Thanks for the great PR! I'm kind of new to Kubernetes so I'm a little confused here but it seems like the readiness probe is going to return a `200 OK` response irrespective of whether the model is loaded or not right? I was under the assumption that the K8's probes check for status code and not necessarily responses? Should we be adding a test to see what it returns when the model is in fact not loaded?
> 
> Regarding your question "it seems like the readiness probe is going to return a `200 OK` response irrespective of whether the model is loaded or not right?", in Kubernetes, when you configure your deployment, you can use startup probes to determinate when a container application has started. Liveness and readiness probes do not start until startup probes succeeds. It allows those probes not to interfere with your application startup. This is particularly useful when you have slow starting containers (for model loading for instance) and to do liveness checks on them. It will avoid them getting killed before they are up and running.

I think the confusion stems from here. It seems that the readiness response incorrectly returns a 200 response (with value "ko", not sure whether it means anything to Kubernetes) even when the model hasn't finished loading yet.

Reply:
> > > @mfournioux Thanks for the great PR! I'm kind of new to Kubernetes so I'm a little confused here but it seems like the readiness probe is going to return a `200 OK` response irrespective of whether the model is loaded or not right? I was under the assumption that the K8's probes check for status code and not necessarily responses? Should we be adding a test to see what it returns when the model is in fact not loaded?
> > 
> > 
> > Regarding your question "it seems like the readiness probe is going to return a `200 OK` response irrespective of whether the model is loaded or not right?", in Kubernetes, when you configure your deployment, you can use startup probes to determinate when a container application has started. Liveness and readiness probes do not start until startup probes succeeds. It allows those probes not to interfere with your application startup. This is particularly useful when you have slow starting containers (for model loading for instance) and to do liveness checks on them. It will avoid them getting killed before they are up and running.
> 
> I think the confusion stems from here. It seems that the readiness response incorrectly returns a 200 response (with value "ko", not sure whether it means anything to Kubernetes) even when the model hasn't finished loading yet.

Ok I see, thanks for the clarification, the readiness probe should not return 200 if not ready, I correct it.

Reply:
I think if `None` is returned from the function, then 200 OK is still returned. You should return an error response explicitly.

Reply:
> I think if `None` is returned from the function, then 200 OK is still returned. You should return an error response (or whatever Kubernetes expects) explicitly.

Indeed, Kubernetes should expect any other code less than 200 and greater or equal to 400 indicates failure.